lib/datadog/core/utils/sequence_numeric.rb

```diff
@@ -0,0 +1,26 @@
+# typed: true
+
+module Datadog
+  module Core
+    module Utils
+      # Generates values from a consistent numeric sequence
+      class SequenceNumeric
```

Comment:
Why was this class created? I don't see much difference from `Sequence`.

Reply:
Based on discussion with @delner, this version of `Sequence` is numeric only - this is done to ensure it is thread-safe (as it is also called by the `Heartbeat` worker)

Reply:
I suggested this as a way of avoiding a block, and simplifying operations to make it thread-safe.

Reply:
Wait: this is not thread safe at all. The increment is not atomic. Here's it breaking in JRuby:

```
ivo.anjo@macieira:~/datadog/third-dd-trace-rb$ be pry
[1] pry(main)> require 'ddtrace'
=> true
[2] pry(main)> sequence_numeric = Datadog::Core::Utils::SequenceNumeric.new
=> #<Datadog::Core::Utils::SequenceNumeric:0x571903ff @current=0, @increment=1, @seed=0>
[3] pry(main)> 10.times.map { Thread.new { 1_000_000.times { sequence_numeric.next } } }.map(&:join)
=> [#<Thread:0x3458be63@(pry):3 dead>,
 #<Thread:0x5cb32140@(pry):3 dead>,
 #<Thread:0x139b3a29@(pry):3 dead>,
 #<Thread:0x62c1259f@(pry):3 dead>,
 #<Thread:0x74267ece@(pry):3 dead>,
 #<Thread:0x3e265811@(pry):3 dead>,
 #<Thread:0x33986be7@(pry):3 dead>,
 #<Thread:0x748fa354@(pry):3 dead>,
 #<Thread:0x78e59b04@(pry):3 dead>,
 #<Thread:0x2e6f1f48@(pry):3 dead>]
[4] pry(main)> sequence_numeric
=> #<Datadog::Core::Utils::SequenceNumeric:0x571903ff @current=2677471, @increment=1, @seed=0>
```

(I had a harder time breaking it in MRI, but I only spent 5 minutes on it. I'm really sure I could break in on MRI as well though')

Reply:
I'd like to make this conversation objectively less productive by suggesting that we should depend on https://rubygems.org/gems/concurrent-ruby at some point in the future. It's pretty much the "informal conclusion" of the discussions around official atomic integer support: https://bugs.ruby-lang.org/issues/12607

Not for this PR.

I'm not sure the telemetry client will ever be used across different threads, is that possible?
If so, I'm happy with a mutex here.

Reply:
(I'll note here that I came in without context. Almost all of dd-trace-rb is **not** thread-safe, so if we can change the code to not need to be executed across multiple threads, that also "solves" the problem -- no concurrency, no concurrency-related problems. But dropping "thread-safe" in a comment brings me immediately around just like saying "bloody mary" 3 times to a mirror.)

Reply:
Thank you for raising this issue! I've added in a mutex now as per @marcotc's suggestion

Reply:
Ruby is notoriously unfriendly for these kinds of things because the standard library is missing a lot of the nice concurrency tools that other languages ship >_>

Reply:
As an alternative to mutex, I tried using a `Fiber` in `SequenceNumeric` like this:
```
    def next
      fiber = Fiber.new do
        next_item = @current
        @current += @increment
        Fiber.yield next_item
      end
      fiber.resume
    end
```

However when comparing its performance to the version with a mutex (running code from Ivo's comment above to test), it seemed to have pretty significant impact (though I'm not sure if I'm using `Fiber` correctly here):
```
# Array.new(10) { Thread.new { 1_000_000.times { sequence_numeric.next } } }.map(&:join)

                                     user     system      total        real
Fiber                           12.925868  24.862574  37.788442 ( 37.755493)
Mutex                            1.054396   0.006688   1.061084 (  1.060875)
```

For now I have removed the mutex, but very open to more suggestions on how to improve this class/its thread-safety 🙏

Reply:
Fibers are pausable code blocks that can return values multiple times.
They will run their code block once and terminate.
You can return multiple times by calling `Fiber.yield val` each time you want to return a value. You could wrap `Fiber.yield val` in a loop for example, making it look like you have an infinite generator.

The issue with Fibers here is that their are not thread-safe: they still execute in one serial pass. Actually, you can't even call `fiber.resume` from multiple threads, you get this error:
```
in `resume': fiber called across threads (FiberError)
```

So if multiple threads are interacting with your Sequence class, you can't use Fibers to generate the value.

Fiber work well when you have to wait for some operation, for example slow IO operations. You can `Fiber.yield` so that the caller can continue, and they can `fiber.resume` later when they are ready to consume the result of the fiber IO operation.

Reply:
I think this question wasn't answered yet, and it's the most important:

> the telemetry client will ever be used across different threads?

Why was thread-safeness a concern in first place?

Like Ivo said, most of our code does not interact with multiple threads at the same time.
If it does, it's normally through very few, well-known places, like when we send traces to the background thread through a [Buffer](https://github.com/DataDog/dd-trace-rb/blob/master/lib/datadog/tracing/buffer.rb).

Reply:
Assuming that we do need thread-safety here (and I think Marco's point is very important), **I can't think of any primitive or tool shipped in the Ruby standard library that would be faster (without significant downsides) that a lock**.

Effectively (that I can think of), there's no single piece of pure Ruby code that you can write to make an increment and write atomic. You need a primitive that goes beyond pure Ruby code into the C level (or possibly into the hardware level).

Furthermore, since we still endeavor to support JRuby customers, the solution should not rely on implementation details of the regular Ruby VM ("MRI"/"CRuby"). (I can think of a few awful hacks otherwise)

Reply:
This conversation may have got a little out of hand.

Given this is a "core" component that could conceivably be used in a multi-threading way, I suggested exploring for an inexpensive way, if any, to make this thread-safe. It's conceivable such a component may be used by telemetry (e.g. emitting errors from multiple threads), or other components in a way that would benefit from thread safety.

So the purpose of the suggestion was to explore, learn, and apply a relevant pattern, if at all reasonable. It's also entirely conceivable that such an application is unreasonable for our needs right now.

If thread safety is cheap to implement/execute, and mutex is the best-worst option, then that's acceptable. It's also acceptable to not do it at all if we can reason that thread safety is unnecessary or unreasonable to implement. I'm fine with deferring to your judgment on this @ivoanjo / @marcotc .

---

Side note:

You could make this mutex behavior "configurable" on the class e.g. `Sequence.new(seed, mutex: true)` to allow for faster, non-thread safe use. Just an idea.

Reply:
Thanks for the clarification. I would suggest not doing any thread-safety stuff until we actually need it.

That's because, in a concrete situation where we're working across threads, there may be a better high-level solution for the need, e.g. if we need to change 3 counters in a row, we could perform the locking in such a way as to encompass all three increments, not just one.

Reply:
Fair enough. Skip on the mutex altogether then.